server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java

```diff
@@ -0,0 +1,511 @@
+        // Only have to make one pass through the list because it is sorted by centroid
+        int closest_idx = 0; // After this loop, (closest_idx, closest_idx + 1) will be the 2 closest buckets
+        double smallest_distance = Double.POSITIVE_INFINITY;
+        for(int i = 0; i < buckets.size() - 1; i++){
+            double new_distance = buckets.get(i+1).centroid - buckets.get(i).centroid; // Positive because buckets is sorted
+            if(new_distance < smallest_distance){
+                closest_idx = i;
+                smallest_distance = new_distance;
+            }
+        }
+
+        mergeTwoBuckets(closest_idx, closest_idx + 1, buckets, reduceContext);
+    }
+
+    /**
+     * Adjust each bucket's min bounds so that no buckets overlap.
+     *
+     * Overlap could occur because when buckets are produced as XContent their minimum values are used as their keys and
+     * their maximum values are not considered.
+     * If two buckets do overlap, then the one with the smaller centroid will contain more values than indicated,
+     * and the one with the larger centroid will contain less than indicated.
+     **/
+    private void adjustMinBoundsIfOverlap(List<Bucket> buckets){
+        for(int i = 0; i < buckets.size(); i++){
+            if(i != 0){
+                Bucket curBucket = buckets.get(i);
+                Bucket prevBucket = buckets.get(i-1);
+                if(curBucket.bounds.min < prevBucket.bounds.max){
+                    // Remove bucket overlap.
+                    // Setting the new bound to the midpoint will typically increase bucket accuracy.
+                    // Otherwise, values from <code>prevBucket</code> could be included in <code>curBucket</code>,
+                    // but no values from <code>curBucket</code> would be in <code>prevBucket</code>.
+                    curBucket.bounds.min = (prevBucket.bounds.max + curBucket.bounds.min) / 2;
+                }
+            }
+        }
+    }
+
+    private boolean mergeTwoBucketsWIthIdenticalMin(List<Bucket> buckets, ReduceContext reduceContext){
+        for (int i = 0; i < buckets.size(); i++) {
+            for(int j = i + 1; j < buckets.size(); j++){
+                if(buckets.get(i).bounds.min == buckets.get(j).bounds.min){
+                    mergeTwoBuckets(i, j, buckets, reduceContext);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Edge case: if two buckets have identical mins, then their centroids are likely close enough to be merged together.
+     * Merging them is necessary in order to remove duplicate keys from the final XContent body.
+     */
+    private void mergeBucketsWidthIdenticalMin(List<Bucket> buckets, ReduceContext reduceContext){
+        // Merge until no more buckets have identical mins.
+        while(mergeTwoBucketsWIthIdenticalMin(buckets, reduceContext));
```

Comment:
We should probably have some kind of counter here that bails after `n` iterations, just in case we hit some worst-case edge-case and it never converges, or takes millions of iterations or something crazy.  Shouldn't happen, but just in case :)